Prevent the unmap length of a negative number

Vicente-Cheng opened the pull request:
#### Which issue(s) this PR fixes:
Issue: https://github.com/longhorn/longhorn/issues/8235

#### What this PR does / why we need it:
Prevent the unmap length of a negative number

#### Special notes for your reviewer:

#### Additional documentation or context

Also, add unittest for unmap operations

derekbit commented on pkg/replica/diff_disk.go:
```diff
@@ -304,10 +308,15 @@ func (d *diffDisk) UnmapAt(unmappableDisks []string, length uint32, offset int64
 		length -= uint32(endSectorOffset)
 	}
-	if length <= 0 {
+	if length == 0 {
```
Is it possible that length is smaller than 0?

shuo-wu replied:
Why cannot this help handle the negative `length` case?

Vicente-Cheng replied:
Because the length is uint32, it overflows when the result is negative.
The original check was fine, but it could not capture the negative case, so I changed it to `==` instead of `<=`

derekbit replied:
OK. Using `<=` rather than `==` is due to that the data type of `length` is unsigned.

derekbit commented on pkg/replica/diff_disk.go:
```diff
@@ -304,10 +308,15 @@ func (d *diffDisk) UnmapAt(unmappableDisks []string, length uint32, offset int64
 
+	// final check
+	if origLength < length {
+		return 0, fmt.Errorf("FATAL, final unmap length(%v) should not be larger than original length(%v)", length, origLength)
```
```suggestion
		return 0, fmt.Errorf("final unmap length(%v) should not be larger than original length(%v)", length, origLength)
```

derekbit commented on pkg/replica/diff_disk.go:
```diff
@@ -304,10 +308,15 @@ func (d *diffDisk) UnmapAt(unmappableDisks []string, length uint32, offset int64
 	}
 
+	// final check
+	if origLength < length {
```
Is it possible?

Vicente-Cheng replied:
No, I thought it was impossible after this patch.
I would like to have this because I am afraid I missed some cases.
We could avoid the unexpected punch hole length and get the error.

WDYT?

derekbit replied:
I'm fine with it. Can you add a comment for the check to avoid any confusion?

Vicente-Cheng replied:
Sure, updated the comment!

shuo-wu replied:
the unexpected punch hole length? I was thinking it handles `length` being overflowed to a large number

Vicente-Cheng replied:
The length being overflowed to a large number means it might be a large number for the punch hole length.
That's why I say the unexpected punch hole length.

This checks here because I want to create some punch hole protection.
The final length is bigger than the original length, which means the extra sector will be punch hole. It caused data corruption.

I update the comment again. Hope that might be more clear.
Thanks!

derekbit commented on pkg/replica/diff_disk.go:
```diff
@@ -293,7 +293,11 @@ func (d *diffDisk) UnmapAt(unmappableDisks []string, length uint32, offset int64
 	if length == 0 {
 		return 0, nil
 	}
+	if length < uint32(d.sectorSize) {
+		return 0, nil
+	}
```
```suggestion
        if length == 0 || length < uint32(d.sectorSize) {
	        return 0, nil
        }
```

Vicente-Cheng replied:
Agree, thanks for the suggestions!

derekbit commented:
```
Running: golangci-lint run --new-from-patch=1064.patch
Running: go fmt
pkg/replica/diff_disk.go
time="2024-03-22T06:55:58Z" level=fatal msg="exit status 1"
```

@Vicente-Cheng Can you check the faiulre?

Vicente-Cheng commented:
> ```
> Running: golangci-lint run --new-from-patch=1064.patch
> Running: go fmt
> pkg/replica/diff_disk.go
> time="2024-03-22T06:55:58Z" level=fatal msg="exit status 1"
> ```
> 
> @Vicente-Cheng Can you check the failure?

Fixed... some indent not align.

Also, remove the `length == 0` because it is covered by `length < uint32(d.sectorSize)` with your suggestions.
(Sorry for the wrong rebase, updated.)

Please help to recheck, thanks!

derekbit reviewed (approved):
LGTM

derekbit commented:
One more question, what's the size of the length that `UnampAt received in the corrupted filesystem?

Vicente-Cheng commented:
> One more question, what's the size of the length that `UnampAt received in the corrupted filesystem?
 I thought the 512 bytes

innobead commented on pkg/replica/replica_test.go:
```diff
@@ -1119,3 +1119,65 @@ func (s *TestSuite) TestUnmapMarkDiskRemoved(c *C) {
+	c.Assert(err, IsNil)
+	c.Assert(ret, Equals, 0)
+
+	// offset = 2048, unmap length = 4096: 1 sector can be unmapped
```
// offset = 2048, unmap length = 8192: 1 sector can be unmapped

Vicente-Cheng replied:
nice catch! Updated!

innobead reviewed (changes requested):
LGTM, just one comment.

innobead commented:
@mergify backport v1.6.x v1.5.x

innobead commented:
cc @DamiaSan fyi
